_posts/en/newsletters/2020-06-10-newsletter.md

```diff
@@ -0,0 +1,311 @@
+
+- **Fee overpayment attack on multi-input segwit transactions:** The
+  transaction fee for a Bitcoin transaction is the difference between
+  the amounts of all the UTXOs that transaction spends and the amounts
+  of all the UTXOs it creates.  Transactions explicitly state the amounts
+  of the UTXOs they create, but the amounts of the UTXOs they spend can
+  only be found by looking at the previous transactions that created
+  those UTXOs.  The spending transaction simply commits to the txid and
+  location of the UTXOs it wants to spend, requiring other software that
+  wants to calculate the fee to either look up each UTXO's previous
+  transaction or to maintain a database of verified UTXO data.
+
+    Hardware wallets don't maintain a UTXO set, so the only trustless
+    way for them to determine the amount of fee paid by a transaction
+    with legacy UTXOs is to get a copy of each legacy UTXO's previous
+    transaction, hash the previous transaction to ensure its txid
+    matches the UTXO reference, and use the now-verified amount of the
+    UTXO to perform the fee calculation.  Since legacy transactions can
+    be almost as large as one megabyte and it's possible for a spending
+    transaction to reference thousands of previous transactions, this
+    process can require the processing of gigabytes of data by
+    resource-constrained hardware wallets.
+
+    One of the several improvements made in [BIP143][] segwit v0 was an
+    attempt to eliminate this burden by having [signatures commit to the
+    amount][bip143 motivation] of the UTXO they were spending.  This
+    meant that any signatures that committed to an incorrect amount
+    would be invalid, which developers of both [Bitcoin Core][towns
+    benefits] and [hardware wallets][palatinus inpatient] believed would
+    allow signers to safely accept amounts from untrusted programs.
+
+    Unfortunately, signing a single UTXO's amount turned out to be
+    insufficient.  In 2017, Greg Sanders [described][sanders attack] how
+    an attacker could send two versions of the same transaction with two
+    segwit UTXOs to a hardware wallet.  In the first case, the attacker
+    would understate the amount of one of the UTXOs being spent; the
+    second case, they'd understate the amount of the second UTXO.
+    Because both transactions understated the amount of one of the
+    UTXOs, the hardware wallet would under-calculate the fee by that
+    same understated amount.  A user would authorize signing the first
+    transaction based on its understated fee.  The attacker would then
+    claim there was some minor problem and the user needs to sign the
+    same transaction again, but would instead get the user to sign the
+    second transaction (which would look identical to the user).
+    Although both transactions produced are individually invalid because
+    they each contain one signature that commits to a wrong amount, each
```

Comment:
I'd remove the part about the invalid signature. The attacker doesn't need to provide a signature at all.

Reply:
I don't understand this comment.

Reply:
Oh, I misunderstood. I didn't understand that both UTXOs are owned by the victim. Perhaps that needs to be explicitly stated for slow people like me?